src/ppx/geometry.cpp

```diff
@@ -13,12 +13,392 @@
+                case grfx::VERTEX_SEMANTIC_TANGENT   : AppendDataToVertexBuffer(pGeom, GetTangentBufferIndex(pGeom), vtx.tangent); break;
+                case grfx::VERTEX_SEMANTIC_BITANGENT : AppendDataToVertexBuffer(pGeom, GetBitangentBufferIndex(pGeom), vtx.bitangent); break;
+                case grfx::VERTEX_SEMANTIC_TEXCOORD  : AppendDataToVertexBuffer(pGeom, GetTexCoordBufferIndex(pGeom), vtx.texCoord); break;
+            }
+            // clang-format on
+        }
+        uint32_t endSize = GetVertexBufferSize(pGeom, kNonPositionBufferIndex);
+
+        uint32_t       bytesWritten            = (endSize - startSize);
+        const uint32_t vertexBufferElementSize = GetVertexBufferElementSize(pGeom, kNonPositionBufferIndex);
+        PPX_ASSERT_MSG(bytesWritten == vertexBufferElementSize, "size of vertex data written does not match buffer's element size");
+        return n;
+    }
+
+    virtual uint32_t AppendVertexData(Geometry* pGeom, const WireMeshVertexData& vtx) override
+    {
+        const uint32_t n = AppendDataToVertexBuffer(pGeom, kPositionBufferIndex, vtx.position);
+        PPX_ASSERT_MSG(n > 0, "position should always available");
+        uint32_t       startSize = GetVertexBufferSize(pGeom, kNonPositionBufferIndex);
+        const uint32_t attrCount = GetVertexBindingAttributeCount(pGeom, kNonPositionBufferIndex);
+        for (uint32_t attrIndex = 0; attrIndex < attrCount; ++attrIndex) {
+            const grfx::VertexSemantic semantic = GetVertexBindingAttributeSematic(pGeom, kNonPositionBufferIndex, attrIndex);
+
+            // clang-format off
+            switch (semantic) {
+                default                              : PPX_ASSERT_MSG(false, "should not have other sematic"); break;
+                case grfx::VERTEX_SEMANTIC_POSITION  : PPX_ASSERT_MSG(false, "position should be in binding 0"); break;
+                case grfx::VERTEX_SEMANTIC_COLOR     : AppendDataToVertexBuffer(pGeom, GetColorBufferIndex(pGeom), vtx.color); break;
+            }
+            // clang-format on
+        }
+        uint32_t endSize = GetVertexBufferSize(pGeom, kNonPositionBufferIndex);
+
+        uint32_t bytesWritten = (endSize - startSize);
+
+        const uint32_t vertexBufferElementSize = GetVertexBufferElementSize(pGeom, kNonPositionBufferIndex);
+        PPX_ASSERT_MSG(bytesWritten == vertexBufferElementSize, "size of vertex data written does not match buffer's element size");
+        return n;
+    }
+
+    virtual uint32_t GetVertexCount(const Geometry* pGeom) override
+    {
+        return GetVertexBufferElementCount(pGeom, GetPositionBufferIndex(pGeom));
+    }
+
+private:
+    const uint32_t kPositionBufferIndex    = 0;
+    const uint32_t kNonPositionBufferIndex = 1;
+};
+
+static VertexDataProcessorPlanar         sVDProcessorPlanar;
+static VertexDataProcessorInterleaved    sVDProcessorInterleaved;
+static VertexDataProcessorPositionPlanar sVDProcessorPositionPlanar;
```

Comment:
Not really an issue with the current design, but if data members ever got added to any of these classes or the base class - it could create problems for multithreaded cases of multiple Geometry objects. Might be useful to have a comment on the classes to warn of this for any future modifications.

Reply:
added some comments at the beginning of this file before the declaration of  `VertexDataProcessorBase`